ld.c ファイルの分割作業

tctsigemura opened the pull request:
modified:   GNUmakefile

tctsigemura committed:
作業の開始を追記

	modified:   GNUmakefile

i20yokota committed:
ld.cの分割途中のファイルです。コンパイルエラーが起きます
	modified:   ld.c
	new file:   rel.c
	new file:   sym.c
	new file:   sym.h
	new file:   util.c
	new file:   util.h

i20yokota committed:
str.c,str.hを追加、他のファイルも変更しました

tctsigemura commented:
コンパイルはできるようになりましたか？

i20yokota committed:
それぞれのファイルを表ごとに分割し、コンパイルエラーを消しました。
実行可能かどうかはまだ確認しておりません。
 Changes to be committed:
	modified:   ld.c
	modified:   rel.c
	new file:   rel.h
	modified:   str.c
	modified:   str.h
	modified:   sym.c
	modified:   sym.h
	modified:   util.c
	modified:   util.h

i20yokota committed:
PCの移行が完了したのでpushします

i20yokota committed:
現状のクラス図です

tctsigemura committed:
ファイルのモードを変更

	modified:   GNUmakefile
	modified:   Makefile
	modified:   classFigure.drawio
	modified:   classFigure.png
	modified:   ld.c
	modified:   rel.c
	modified:   rel.h
	modified:   str.c
	modified:   str.h
	modified:   sym.c
	modified:   sym.h
	modified:   util.c
	modified:   util.h

tctsigemura committed:
図の左右幅を小さくした

	modified:   "\343\202\257\343\203\251\343\202\271\345\233\2630.asta"

i20yokota committed:
9月5日の進捗
	modified:   LD--/ld.c
	modified:   LD--/rel.c
	modified:   LD--/str.c
	modified:   LD--/sym.c
	modified:   LD--/util.c
	modified:   LD--/util.h

i20yokota committed:
Merge branch 'split' of https://github.com/newutil/Util-- into split

i20yokota committed:
9月5日時点の進捗
	modified:   "LD--/\343\202\257\343\203\251\343\202\271\345\233\2630.asta"

tctsigemura committed:
重村の提案

	modified:   LD--/ld.c
	modified:   LD--/rel.c
	modified:   LD--/rel.h
	modified:   LD--/util.c
	modified:   LD--/util.h

i20yokota committed:
9月13日の進捗
	modified:   LD--/ld.c
	modified:   LD--/rel.c
	modified:   LD--/str.c
	modified:   LD--/sym.c
	modified:   LD--/util.c
	new file:   "LD--/\343\202\257\343\203\251\343\202\271\345\233\2630.asta.lock"

tctsigemura committed:
間違って追加された lock ファイルを削除

	deleted:    "LD--/\343\202\257\343\203\251\343\202\271\345\233\2630.asta.lock"

tctsigemura committed:
relクラスの詳細化

	modified:   "LD--/\343\202\257\343\203\251\343\202\271\345\233\2630.asta"

i20yokota committed:
9月19日の進捗
	modified:   ld.c
	modified:   rel.c
	modified:   str.c
	modified:   sym.c
	modified:   util.c
	modified:   util.h
	modified:   "\343\202\257\343\203\251\343\202\271\345\233\2630.asta"

i20yokota committed:
Merge branch 'split' of https://github.com/newutil/Util-- into split

	LD--/クラス図0.asta

i20yokota committed:
9月26日の進捗
	modified:   ld.c
	modified:   rel.c
	modified:   str.c
	modified:   sym.c
	modified:   util.c
	modified:   util.h
	modified:   "\343\202\257\343\203\251\343\202\271\345\233\2630.asta"

i20yokota committed:
9月26日　報告中の進捗
	modified:   ld.c
	modified:   sym.c
	modified:   sym.h
	modified:   util.h
	modified:   "\343\202\257\343\203\251\343\202\271\345\233\2630.asta"

i20yokota committed:
コンパイルエラーを解決しました。
	modified:   ld--
	modified:   ld.c
	modified:   rel.c
	modified:   rel.h
	modified:   str.c
	modified:   str.h
	modified:   sym.c
	modified:   sym.h
	modified:   util.c
	modified:   util.h
	modified:   "\343\202\257\343\203\251\343\202\271\345\233\2630.asta"

tctsigemura committed:
LD--のテストができるように古いテスト環境を更新

	modified:   LD--/Test/Makefile
	modified:   LD--/Test/a.cmm
	deleted:    LD--/Test/a.s
	modified:   LD--/Test/b.cmm
	modified:   LD--/Test/c.cmm
	deleted:    LD--/ld--

i20yokota committed:
クラス図の整理とプログラムのデバッグを行いました
	new file:   ld--
	modified:   ld.c
	modified:   rel.c
	modified:   rel.h
	modified:   str.c
	modified:   str.h
	modified:   sym.c
	modified:   sym.h
	modified:   util.c
	modified:   util.h
	modified:   "\343\202\257\343\203\251\343\202\271\345\233\2630.asta"

i20yokota committed:
.aファイルの中身を全て読み込むようutilを変更し、テスト用アーカイブファイルtArc.aで動作を確認しました。
	modified:   Test/a.cmm
	modified:   Test/b.cmm
	modified:   Test/c.cmm
	new file:   Test/tArc.a
	new file:   Test/testArc.hmm
	new file:   Test/testArc2.hmm
	modified:   ld.c
	modified:   util.c
	modified:   util.h
	modified:   "\343\202\257\343\203\251\343\202\271\345\233\2630.asta"

tctsigemura commented on LD--/util.c:
```diff
@@ -30,34 +42,104 @@ void tblError(char *str, int idx, int size) {  //表がパンクした時のエ
+
+  if((curAFile[i-2]!='.'||curAFile[i-1]!='o')){ //.oファイル以外があった時
+    fError("アーカイブファイル内に扱えないファイルが存在\n");
+  }
```
この処理は省略しましょう．
1. 後でマジックナンバーで確認している．
2. これだけのために curAFile を準備するのもおおがかり．

tctsigemura commented on LD--/util.c:
```diff
@@ -30,34 +42,104 @@ void tblError(char *str, int idx, int size) {  //表がパンクした時のエ
+  c = getW();
+  printf("debug: masic number：%x\n",c);  //デバッグ用
+  if (c!=A_MAGIC)              //   マジックナンバー
+    fError("扱えないアーカイブファイル");
```
カッコで囲ってください．
他にも（重村が書いた部分も含めて）カッコが無いところがあったら囲ってください．
```
  if (c!=A_MAGIC) {             //   マジックナンバー
    fError("扱えないアーカイブファイル");
  }
```

tctsigemura commented on LD--/util.c:
```diff
@@ -30,34 +42,104 @@ void tblError(char *str, int idx, int size) {  //表がパンクした時のエ
+
+    printf("debug: This file \"%s\" is archive file.\n",curFile);  //デバッグ用
+
+    int i=0;
+    int *c;
+    int aLen = strlen(arcCheck);
+    while(i<aLen){
+      c[i] = getB();
+      printf("%c",(char)c[i]);  //デバッグ用
+      if(c[i] != arcCheck[i])             //想定しているファイル形式でなければ
+        fError("アーカイブファイルではない.aファイル\n");   //エラー終了
+      i = i + 1;
+    }; 
```
1. 変数```c```はポインタを確保しただけで配列本体が確保されていません．バグです．
2. そもそも，```c```配列は不要だと思います．配列に読み込んでいるが後で使っていない．

こんな感じにしたらどうでしょうか？
```
int aLen = strlen(arcCheck);
for (int i=0; i<aLen; i=i+1){
  int c = getB();
  printf("%c",(char)c[i]);  //デバッグ用
  if(c != arcCheck[i]) {            //想定しているファイル形式でなければ
    fError("アーカイブファイルではない.aファイル\n");   //エラー終了
  }
}; 
```

tctsigemura replied:
この部分，```checkArc()```みたいな関数に切り出した方が良いと思います．
（```xOpenIn()```の本質的な処理ではないので，```xOpenIn()```の中に直接書いてあると理解しにくい．）
また，ここでしか使用しない```"!<arch>\n"```は関数内に閉じ込める．
```
void arcCheck() {
  arcStr="!<arch>\n";                 // アーカイブファイルの印
  int aLen = strlen(arcStr);
  for (int i=0; i<aLen; i=i+1){
    int c = getB();
    printf("%c",(char)c[i]);  //デバッグ用
    if(c != arcCheck[i]) {            //想定しているファイル形式でなければ
      fError("アーカイブファイルではない.aファイル\n");   //エラー終了
    }
  }; 
}

tctsigemura replied:
```char *arcStr="!<arch>\n";```の間違いです．

tctsigemura commented on LD--/util.c:
```diff
@@ -30,34 +42,104 @@ void tblError(char *str, int idx, int size) {  //表がパンクした時のエ
+    readArchive();                        //一つ目のライブラリ関数の情報を読む
+  }
+  else printf("debug: This file \"%s\" is NOT archive file.\n",curFile);  //デバッグ用
+
```
```isArchive```は```nextFile()```中ではなく，ここで```false```にしたほうがわかりやすいと思います．
1. ```true```にするところの近くにあれば，どこで```false```になるか捜さなくてよい．
2. 「```.a```ファイルをオープン時にどちらか判定する．」と考えるとわかりやすい．

```
  isArchive = true;                      //アーカイブのフラグを立て、
  readArchive();                        //一つ目のライブラリ関数の情報を読む
}  else {
  isArchive = false;
  printf("debug: This file \"%s\" is NOT archive file.\n",curFile);  //デバッグ用
}
```

tctsigemura commented on LD--/util.c:
```diff
@@ -30,34 +42,104 @@ void tblError(char *str, int idx, int size) {  //表がパンクした時のエ
 
-void xSeekIn(int offset) {                   // エラーチェック付きの SEEK ルーチン
-  if ((offset&1)!=0 || fseek(in, (long)offset, SEEK_SET)!=0)
+void xSeekIn(int offset) {  // 入力ファイル用エラーチェック付きの SEEK ルーチン
+  
+  if(isArchive)                   //アーカイブファイルの場合
+    offset = offset + cFileHead;  //ファイルの内容の先頭からSEEKする
+  else if ( (offset&1)!=0){       //アーカイブファイルではない場合
+    fError("file format");        //奇数アドレスへfSeekはエラーを返す
+  }
+  
+  if ( fseek(in, (long)offset, SEEK_SET)!=0)
     fError("file format");
 }
```
```offset```はアーカイブファイ中でも```.o```の先頭からの相対位置ですよね．
なら，どちらでも奇数アドレスへのシークは何か異常が発生している．
次のようにしたらどうでしょうか．
```
void xSeekIn(int offset) {  // 入力ファイル用エラーチェック付きの SEEK ルーチン
  int realOffs = offset;
  if(isArchive) {                  //アーカイブファイルの場合
    realOffs = offset + cFileHead;  //ファイルの内容の先頭からSEEKする
  }
  if ((offset&1)!=0 || fseek(in, (long)realOffs, SEEK_SET)!=0) {
    fError("file format");
  }
}
```

i20yokota replied:
nextFile関数で使っているxSeek関数を、通常のfseek関数に置き換えるとそれでうまく動くかと思います。

前回エラーが起きていた原因はおそらく、util.cの最後尾に書かれているnextFile関数です。
nextFile関数がアーカイブファイルの内容を全て読み込めているか調べるために、
読み込み中.oファイルの最後へxSeekを用いて移動していました。
そのため移動先が奇数アドレスになることがあり、実行時エラーが発生してしまっていたのだと思います。

tctsigemura replied:
そうですね．nextFile()はfseek()を使用すべきですね．

tctsigemura reviewed (commented):
一旦ここまでで，送ります．

tctsigemura commented on LD--/util.c:
```diff
@@ -0,0 +1,180 @@
+  return (x1 << 8) | x2;
+}
+
+boolean nextFile(){      //現在読み込み中のアーカイブファイルの中で、
+                         //次のファイルがあるかどうかを調べる
+  if(!isArchive)
+    return false;         //アーカイブファイルではないときはfalseで終了
+
+  xSeekIn(cFileLen);     //ファイルの内容の最後までseek
+
+  int x = getB();
+  if(x==EOF){           //ファイルが最後まで読み込めている場合
+    isArchive = false;  //isArchiveフラグをfalseにする 
+    printf("debug: nextFile() returned true.\n\tArchive file is already read all.\n"); //デバッグ用
+    return false;
+  }else{                      //まだファイルがあるなら
+
+    printf("debug: nextFile() returned false.\n\tArchive file is not read all yet.\n"); //デバッグ用
+
+      xSeekIn(cFileLen); //読んだ分戻して
+      readArchive();          //次のファイルを読む
+
+      return true;    //trueを返す
+  }
+}
```
少し清書してこんな感じ．変数 ```x```は不要です．
```
boolean nextFile(){      //現在読み込み中のアーカイブファイルの中で、
                         //次のファイルがあるかどうかを調べる
  if (!isArchive) {
    return false;         //アーカイブファイルではないときはfalseで終了
  }

  xSeekIn(cFileLen);     //ファイルの内容の最後までseek
  if (getW()==EOF){           //ファイルが最後まで読み込めている場合
    printf("debug: nextFile() returned true.\n\tArchive file is already read all.\n"); //デバッグ用
    return false;
  }

  printf("debug: nextFile() returned false.\n\tArchive file is not read all yet.\n"); //デバッグ用

  xSeekIn(cFileLen); //読んだ分戻して
  readArchive();          //次のファイルを読む

  return true;    //trueを返す
}
```

tctsigemura commented on LD--/util.h:
```diff
@@ -0,0 +1,29 @@
+void fcloseIn();    
+void fcloseOut();
+void xSeekIn(int offset);                       // エラーチェック付きの SEEK ルーチン
+void xSeekOut(int offset);
+void putW(int x);                               // 1ワード出力ルーチン
+int getW();                                     // 1ワード入力ルーチン
+boolean nextFile();
```
```nextFile()```は，util.c 以外から呼ばれることは無いと思います．
呼ばれないなら，ここに書く必要はありません．

i20yokota replied:
nextFile()はmain関数からの呼び出しがあるので記述しています。

入力ファイルから情報を読み込む際、
読み込み中のアーカイブファイルから次の.oファイルを読み込むか、
ファイルを閉じて次のファイルを読み込むかの判断をnextFile関数に行わせています。

tctsigemura replied:
そうですね．
ld.c からの呼び出しのために存在する関数ですね．

tctsigemura reviewed (commented):
util.c util.h 両方見終わりました．

i20yokota committed:
utilを少し変更したのち、正常に動作していることを確認しました。
	modified:   util.c

tctsigemura commented on LD--/str.c:
```diff
@@ -0,0 +1,92 @@
+  return strIdx;
+}
+
+boolean isStrLocal(int i){                      //文字列表の値がローカルなら正を返す
+  if(strTbl[i] == '.')
+    return true;
+  return false;
+}
```
ここは，次のように書くと良いです．
```
boolean isStrLocal(int i) {
  return strTbl[i]=='.';
}
```

i20yokota replied:
変更しておきます

tctsigemura commented on LD--/str.c:
```diff
@@ -0,0 +1,92 @@
+  return true;
+}
+
+void putStr(FILE* fp,int n) {               // 文字列表の文字列[n]を表示する
+  if (n>0x3fff || n>=strIdx) error("putStr:バグ");
+  while (strTbl[n]!='\0') {
+    putc(strTbl[n],fp);
+    n = n + 1;
+  }
+}
```
```0x3fff```は何の値ですか？

i20yokota replied:
シンボルテーブルの文字列表参照アドレスの範囲が0x3FFFで、
その範囲のアドレスだけマスクした値で、putStrを実行しているようです。
なので、それ以上の値があればバグが発生している、という判定だと思います。
この辺りは分割前から記入されていた部分です。

tctsigemura replied:
マスク忘れのバグを実行時に発見するための仕組みだったようです．

tctsigemura commented on LD--/str.c:
```diff
@@ -0,0 +1,92 @@
+  }
+}
+
+void packStrTbl(int idxI,int len){   //  文字列表から統合した綴りを削除する
+  for (int k=idxI; k<strIdx-len; k=k+1)
+	  strTbl[k] = strTbl[k+len];        //     文字列を前につめる
+	strIdx = strIdx - len;              //   文字列表を縮小
+}
```
TAB文字が入っているので，他の環境で見るとインデントが崩れます．
<img width="565" alt="スクリーンショット 2024-11-12 14 16 10" src="https://github.com/user-attachments/assets/14a51feb-d89c-427c-917d-5a6383d5faea">

i20yokota replied:
修正しました

i20yokota replied:
以前のコメントについても確認したため、今から修正をしていきます。

tctsigemura reviewed (commented):
str.c のチェック中です．
一旦，送ります．

i20yokota committed:
コードを修正しました
	modified:   ld.c
	modified:   rel.c
	modified:   str.c
	modified:   sym.c
	modified:   util.c

tctsigemura commented on LD--/sym.c:
```diff
@@ -0,0 +1,168 @@
+  return bssSize;
+}
+
+void updateSymStrx(int curIdx, int changeIdx, int len) { //文字列表の統合に合わせて
+                                                         //名前表のアドレスを調整する
+  for(int i=0; i<symIdx; i=i+1) {
+    int idxI = symTbl[i].strx;
+    if(idxI == changeIdx) {                 //統合した文字列を指しているならば
+      symTbl[i].strx = curIdx;              //以前からある方に合わせる
+    } else if(idxI >= changeIdx) {          //前に詰めた部分にあるものは
+      symTbl[i].strx = idxI - len;          //位置調整 
+    }
+  }
+
+}
```
111行の条件は ```==``` の場合は 109行ですでに取り扱われているので
```>=``` ではなく ```>``` でよいですね．
```
    } else if(idxI > changeIdx) {
```

tctsigemura replied:
２つ目は sym.c の 111 行のことです．

tctsigemura commented on LD--/str.c:
```diff
@@ -0,0 +1,91 @@
+  }
+}
+
+void mergeStrTbl(int strIdxB) {  // 文字列表に新しく追加した綴りに
+                                              // 重複があれば統合する              
+  int idxI = strIdxB;     // 新しい文字列               
+  int idxJ = 0;           // 元からある文字列
+
+  while(idxI < strIdx) {                      // 追加された文字列について
+    idxJ = 0;
+    while(idxJ < strIdxB) {                   // 以前からある文字列と比較
+      if(cmpStr(idxI, idxJ)) {                // 同じ綴が見つかったら
+        int len = strLen(idxI);
+        updateSymStrx(idxJ, idxI, len);       // 名前表のアドレスを調整して
+        packStrTbl(idxI,len);                 // 文字列表から統合した綴りを削除
+        break;
+      }                                       // 違う文字列の場合は
+      idxJ = idxJ + strLen(idxJ);             // 次の文字列を準備
+    }
+    idxI = idxI + strLen(idxI);               // 追加された文字列を全てチェック      
+  }
+}
```
アルゴリズムにバグはないと思いますが，短く清書しました．
for文を使った方がよいとおもいます．
```
// 文字列表に新規追加の綴に，以前からの綴と重複があれば統合
void mergeStrTbl(int strIdxB) {
  // 追加された全ての綴について
  for (int idxl=strIdxB; idxI<strIdx; idxl=idxl+strLen(idxl)) {
    // 以前からある全ての綴と比較
    for (int idxJ=0; idxJ<strIdxB; idxJ=idxJ+strLen(idxJ) {
      if(cmpStr(idxI, idxJ)) {                // 同じ綴が見つかったら
        int len = strLen(idxI);
        updateSymStrx(idxJ, idxI, len);       // 名前表のアドレスを調整して
        packStrTbl(idxI,len);                 // 文字列表から統合した綴りを削除
        break;                                // 同じ綴は複数存在しない
      }
    }
  }
}
```

tctsigemura reviewed (commented):
str.c 最後まで目を通しました．

tctsigemura commented on LD--/sym.c:
```diff
@@ -0,0 +1,168 @@
+    error("名前表の参照先がおかしい");    // 存在しない番地
+  }
+  return symTbl[index];
+}
```
```SYM_SIZ``` ではなく，```symIdx``` と比較したほうが良いのでは？

i20yokota committed:
tabをスペースに変換してインデントを揃え、str.cとsym.cに変更を加えました
	modified:   ld.c
	modified:   rel.c
	modified:   str.c
	modified:   sym.c

tctsigemura reviewed (commented):
コメントを送り忘れていました．
rel.c 等，他のテーブルでも同じことが言えないでしょうか．

tctsigemura commented on LD--/sym.c:
```diff
@@ -0,0 +1,168 @@
+  putStr(stdout,symTbl[symx].strx);
+}
+
+void packSymTbl() {                         // 名前表の不要エントリーを削除
+  int i = 0;
+  while (i<symIdx) {                        // 全てのエントリーについて
+    if (symTbl[i].type==SYMPTR) {           // PTRなら以下のように削除する
+      updateRelSymx(i);                     // 再配置表のインデクスを調整する
+      for (int j=i; j<symIdx-1; j=j+1) {    // 名前表を前につめる
+        symTbl[j].strx = symTbl[j+1].strx;
+        symTbl[j].type = symTbl[j+1].type;
+        symTbl[j].val  = symTbl[j+1].val;
+      }
+      symIdx = symIdx - 1;                  // 名前表を縮小する
+    } else {
+      i = i + 1;                            // PTR以外なら進める
+    }
+  }
+}
```
もとものld.cですでにそうなっていたかも知れませんが．．．
159から161行は構造体全体をコピーすれば良いので．
次の１行でよいです．
```
  symTbl[j] = symTbl[j+1];
```

tctsigemura commented on LD--/rel.c:
```diff
@@ -0,0 +1,106 @@
+  readRelTbl(offs,cDrSize,symBase,segBase);
+}
+
+void updateRelSymx(int ptrIdx) {
+  for (int j=0; j<relIdx; j=j+1) {          // 再配置情報全てについて
+    if (relTbl[j].symx>=ptrIdx) {           // 名前表の削除位置より後ろを
+      relTbl[j].symx=relTbl[j].symx-1;      // 参照しているインデクスを調整
+    }
+  }
+}
```
73行，コメント通りなら ```>=``` ではなく ```>``` ですね．
コメントが正しいと思います．

tctsigemura reviewed (commented):
もう少し見ました．

tctsigemura commented on LD--/util.c:
```diff
@@ -0,0 +1,187 @@
+  exit(1);
+}
+
+void checkArc() {
+  char *arcStr = "!<arch>\n";       //アーカイブファイル判定用
+  int i=0;
+  int c;
+  int aLen = strlen(arcStr);
+  while(i<aLen) {
+    c = getB();
+    if(c != arcStr[i]) {             //想定しているファイル形式でなければ
+      fError("アーカイブファイルではない.aファイル\n");   //エラー終了
+    }
+    i = i + 1;
+  }; 
+  printf("debug: File header OK.\n");  //デバッグ用
+
+}
```
昨日，コメントしたプログラムのように，
1. ```for```文にしたほうが良い
2. 変数```c```は```for```ループ内のローカル変数にした方が良いです．

コンパクトに書いたほうが読む人の労力が軽減されます．
また，変数などの有効範囲は狭いほうがバグの原因になりにくいです．

tctsigemura commented on LD--/classFigure.drawio:
使用していないファイルだったら消してください．
もしも，再度必要になったら git の機能で復活できます．
ファイルを消すのは次の操作です．
$ git rm ファイル名

tctsigemura replied:
png ファイルもです．
後，astahのファイルのファイル名は英語にした方がよいかな．
% git mv クラス図0.asta ClassDiagram.asta

tctsigemura commented on LD--/str.c:
```diff
@@ -0,0 +1,105 @@
+    }
+  }
+}
+// void mergeStrTbl(int strIdxB) {  // 文字列表に新しく追加した綴りに
+//                                               // 重複があれば統合する              
+//   int idxI = strIdxB;     // 新しい文字列               
+//   int idxJ = 0;           // 元からある文字列
+
+//   while(idxI < strIdx) {                      // 追加された文字列について
+//     idxJ = 0;
+//     while(idxJ < strIdxB) {                   // 以前からある文字列と比較
+//       if(cmpStr(idxI, idxJ)) {                // 同じ綴が見つかったら
+//         int len = strLen(idxI);
+//         updateSymStrx(idxJ, idxI, len);       // 名前表のアドレスを調整して
+//         packStrTbl(idxI,len);                 // 文字列表から統合した綴りを削除
+//         break;
+//       }                                       // 違う文字列の場合は
+//       idxJ = idxJ + strLen(idxJ);             // 次の文字列を準備
+//     }
+//     idxI = idxI + strLen(idxI);               // 追加された文字列を全てチェック      
+//   }
+// }
```
ここも，消してください．
必要なら git の機能で過去の記述が確認できます．

tctsigemura commented on LD--/sym.c:
```diff
@@ -0,0 +1,168 @@
+}
+
+
+void readSymTbl(int offs, int sSize,int textBase,int dataBase) {      // 名前表の読み込み
+
```
ここだけではありませんが，１行，半角80文字以内にしてください．
（全角は半角2文字換算）
ソースコードをプリンタに印刷したりする時，1行，最大80文字が標準です．

tctsigemura reviewed (commented):
もう少し見ました．

i20yokota committed:
ご指摘を元に変更を加えました。
	deleted:    classFigure.drawio
	deleted:    classFigure.png
	modified:   ld.c
	modified:   rel.c
	modified:   str.c
	modified:   sym.c
	modified:   util.c

i20yokota committed:
少し修正を加えました
	modified:   ld.c
	modified:   rel.c
	modified:   str.c
	modified:   sym.c
	modified:   util.c

i20yokota committed:
必要なライブラリ関数のみ取得する機能を実装してみました
	modified:   ClassDiagram.asta
	modified:   ld.c
	modified:   str.c
	modified:   str.h
	modified:   sym.c
	modified:   sym.h
	modified:   util.c
	modified:   util.h

i20yokota committed:
テスト用のアーカイブファイルです
	modified:   tArc.a

i20yokota committed:
クラス図の表記揺れを修正し、main関数の動作を明示しました
	modified:   ClassDiagram.asta

i20yokota commented:
size--を用いて出力したテストファイルの情報を見たところ、結合されたライブラリ関数のサイズが出力ファイルのヘッダに反映されていない問題があることがわかりました。現在は解決しました

i20yokota committed:
ライブラリ関数のサイズが反映されない問題を解決しました
	modified:   ld.c
	modified:   util.c
